src/assisted_installer_controller/assisted_installer_controller.go

```diff
@@ -246,6 +246,98 @@ func (c controller) UpdateBMHs(wg *sync.WaitGroup) {
 	}
 }
 
+func (c controller) UpdateWorkerLinkage(wg *sync.WaitGroup) {
+	defer wg.Done()
+	log := c.log.WithField("Job", "WorkerLinker")
+	for {
+		time.Sleep(GeneralWaitInterval)
+
+		exists, err := c.kc.IsMetalProvisioningExists()
+		if err != nil {
+			continue
+		}
+		if err == nil && exists {
+			log.Infof("Provisioning CR exists, no need to update BMHs")
+			return
+		}
+
+		machines, err := c.kc.ListMachines()
+		if err != nil {
+			log.WithError(err).Errorf("Failed to get nodes")
+			continue
+		}
+
+		bmhs, err := c.kc.ListBMHs()
+		if err != nil {
+			log.WithError(err).Errorf("Failed to BMH hosts")
+			continue
+		}
+
+		allUpdated := c.updateWorkerLinkage(bmhs, machines)
+		if allUpdated {
+			log.Infof("Updated all the worker links, finished successfully")
+			return
+		}
+	}
+}
+
+func (c controller) updateWorkerLinkage(bmhList metal3v1alpha1.BareMetalHostList, machineList *mapiv1beta1.MachineList) bool {
+	allUpdated := true
+	log := c.log.WithField("Job", "WorkerLinker")
+
+	assignedMachines := []string{}
+	for _, bmh := range bmhList.Items {
+		if bmh.Spec.ConsumerRef != nil {
+			assignedMachines = append(assignedMachines, bmh.Spec.ConsumerRef.Name)
+		}
+	}
+
+	for i := range bmhList.Items {
+		bmh := bmhList.Items[i]
+		c.log.Infof("Checking bmh %s", bmh.Name)
+
+		if bmh.Spec.ConsumerRef != nil {
+			continue
+		}
+		allUpdated = false
+
+		machine, err := c.findUnassignedWorkerMachine(machineList, assignedMachines)
+		if err != nil {
+			log.WithError(err).Errorf("Failed to find an unassigned worker machine")
+			continue
+		}
+
+		bmh.Spec.ConsumerRef = &v1.ObjectReference{
+			APIVersion: machine.APIVersion,
+			Kind:       machine.Kind,
+			Namespace:  machine.Namespace,
+			Name:       machine.Name,
+		}
+
+		log.Infof("updating bmh consumerRef")
+		err = c.kc.UpdateBMH(&bmh)
+		if err != nil {
+			log.WithError(err).Errorf("Failed to remove status annotation from BMH %s", bmh.Name)
+		} else {
+			assignedMachines = append(assignedMachines, machine.Name)
```

Comment:
I think that you can simply delete an Item from MachineList , and not to managed an additional list (assignedMachines).